### DISCUSSION TITLE
Pretty-printing Hy forms

### DISCUSSION THREAD

**Discussion by mcejp** [hylang/hy#2478] [Q&A] [answered]
As far as I understand, Hy models like `Expression` can be stringified with `hy.repr` and they will correctly round-trip with `hy.read`. However, the output is not "pretty" in the sense of using newlines and indentation for nested structures.

On the other hand, Hyrule has a [pretty-printing module](https://hyrule.readthedocs.io/en/master/#module-hyrule.hypprint), however, it does not seem to be aware of `hy.models` classes.

Is there a solution? Is this something that would be worth adding to Hyrule's PrettyPrinter, or am I misunderstanding its purpose?

**Comment by Kodiologist** [comment, marked as answer]
There are probably a lot of sensible extensions one could make to `hyrule.pprint`, yeah. It hasn't gotten a lot of attention since it was first written a while ago.
